### PR TITLE
Add unit tests for Entities

### PR DESCRIPTION
**Pull request by tzhelyazkova**

### CODE REVIEW

File: tests/Unit/Entity/BannerSelection/CampaignTest.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+			[],
+			new TrafficImpressionThreshold( 10, 10 )
+		);
+		$this->assertFalse(
+			$campaign->isInActiveDateRange( new \DateTime( '2018-09-22' ) ),
+			'date given is before the start of the campaign'
+		);
+	}
+
+	public function testGivenBucketId_returnsBucket() {
```
Comment (gbirke) [inline tests/Unit/Entity/BannerSelection/CampaignTest.php]: See https://github.com/wmde/banner-server/pull/15#discussion_r227886020

File: tests/Unit/Entity/BannerSelection/CampaignTest.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+			[],
+			new TrafficImpressionThreshold( 10, 10 )
+		);
+		$this->assertEquals(
+			$campaign->selectBucket( 'C18_WMDE_Test_var', '$this->returnRandomBucket' ),
+			'C18_WMDE_Test_var'
+		);
+	}
+
+	public function testGivenNoBucketId_returnsRandomBucket() {
```
Comment (gbirke) [inline tests/Unit/Entity/BannerSelection/CampaignTest.php]: `testGivenInvalidBucketId_bucketIsChosenByFallbackFunction`

File: tests/Unit/Entity/BannerSelection/CampaignTest.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		$campaign = new Campaign(
+			'C18_WMDE_Test',
+			new \DateTime( '2018-10-01' ),
+			new \DateTime( '2018-10-31' ),
+			[],
+			new TrafficImpressionThreshold( 10, 10 )
+		);
+		$this->assertEquals( $campaign->selectBucket( null, '$this->returnRandomBucket' ), '' );
```
Comment (gbirke) [inline tests/Unit/Entity/BannerSelection/CampaignTest.php]: An anonymous function would be sufficient. 

If you want to use a public static function from this test class, pass `[CampaignTest::class, 'returnRandomBucket']` instead. Unfortunately, we have to reference methods by string, but for class names, we can always use the `::class`. This helps later, when we rename the class in PHPStiorm refactoring.

File: tests/Unit/Entity/BannerSelection/CampaignTest.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+	public static function returnRandomBucket() {
+		return '';
+	}
+
+	public function testGivenTimeOutOfDateRange_campaignIsNotActive() {
+		$campaign = new Campaign(
+			'C18_WMDE_Test',
+			new \DateTime( '2018-10-01' ),
+			new \DateTime( '2018-10-31' ),
```
Comment (gbirke) [inline tests/Unit/Entity/BannerSelection/CampaignTest.php]: This test is perfectly valid, but reminds me that when you're writing the parser you should either require exact dates with times (and make sure that the timezone is set) or choose different default times for start and end date. A day always starts at 0:00:00 and always ends at 23:59:59.

**Comment by gbirke** [comment]
Just to give you an idea of possible other PHPUnit test naming schemes:

`testGivenSomeStateThenIExpectSomeResult`
`test_given_some_state_then_i_expect_some_result`

I like the `snake_case` one. We'd have to make some exceptions to the phpcs config for the test.

File: tests/Unit/Entity/BannerSelection/CampaignTest.php (hunk 2 above)
Comment (tzhelyazkova) [reply tests/Unit/Entity/BannerSelection/CampaignTest.php]: hm, that makes me think we need one more test, because you can pass an invalid `bucketId` and you can purposefully pass null which I wouldn't consider under the term invalid.

File: src/Entity/BannerSelection/Bucket.php [hunk 5]
```diff
@@ -10,30 +10,35 @@
...
 
 	/**
 	 * @var Banner[]
 	 */
 	private $banners = [];
 
-	public function __construct( string $identifier, Banner $mainBanner, Banner ...$otherBanners ) {
+	public function __construct( string $identifier, Banner $mainBanner, array $otherBanners ) {
 		$this->identifier = $identifier;
+		$this->mainBanner = $mainBanner;
```
[the remark this answers is not in the document]
Comment (tzhelyazkova) [reply src/Entity/BannerSelection/Bucket.php]: I introduced it because of `getBanner` method but then I refactored it and forgot to remove this from the constructor.
Thanks!

File: src/Entity/BannerSelection/Bucket.php [hunk 6]
```diff
@@ -10,30 +10,35 @@
...
 	private $identifier;
+	private $mainBanner;
 
 	/**
 	 * @var Banner[]
 	 */
 	private $banners = [];
 
-	public function __construct( string $identifier, Banner $mainBanner, Banner ...$otherBanners ) {
+	public function __construct( string $identifier, Banner $mainBanner, array $otherBanners ) {
```
[the remark this answers is not in the document]
Comment (tzhelyazkova) [reply src/Entity/BannerSelection/Bucket.php]: I removed it because `$buckets` held the main banner in index 0 and then in index 1 there was an array of other banners. I thought that's not right and made it "flat" array this way.
